core/templates/pages/contributor-dashboard-page/modal-templates/translation-modal.controller.ts

```diff
@@ -100,6 +146,94 @@ angular.module('oppia').controller('TranslationModalController', [
       $scope.activeWrittenTranslation.html = '';
     };
 
+    $scope.getElementTexts = function(elements, type) {
+      const textWrapperLength = 6;
+      const attributes = Array.from(elements, function(element: HTMLElement) {
+        if (element.localName === 'oppia-noninteractive-image') {
+          const attribute = element.attributes[type].value;
+          return attribute.substring(
+            textWrapperLength, attribute.length - textWrapperLength);
+        }
+      });
+      return attributes.filter(attribute => attribute);
+    };
+
+    $scope.copiedAllElements = function(
+        originalElements, translatedElements) {
+      const hasMatchingTranslatedElement = (element) => (
+        translatedElements.includes(element));
+      return originalElements.every(hasMatchingTranslatedElement);
+    };
+
+    $scope.hasSomeDuplicateElements = function(
+        originalElements, translatedElements) {
+      if (originalElements.length === 0) {
+        return false;
+      }
+      const hasMatchingTranslatedElement = (element) => (
+        translatedElements.includes(element) && originalElements.length > 0);
+      return originalElements.some(hasMatchingTranslatedElement);
+    };
+
+    $scope.getTexts = function(rawText) {
+      const foundImageFilePaths = $scope.getElementTexts(
+        rawText, 'filepath-with-value');
+      const foundImageAlts = $scope.getElementTexts(
+        rawText, 'alt-with-value');
+      const foundImageDescriptions = $scope.getElementTexts(
+        rawText, 'caption-with-value');
+
+      return {
+        foundImageFilePaths,
+        foundImageAlts,
+        foundImageDescriptions
+      };
+    };
+
+    $scope.isTranslationCompleted = function(
+        originalElements: [HTMLElement], translatedElements: [HTMLElement]) {
+      const originalHtmlElements = Array.from(
+        originalElements, element => element.nodeName);
+      const translatedHtmlElements = Array.from(
+        translatedElements, element => element.nodeName);
+
+      for (const originalElement of originalHtmlElements) {
+        if (
+          !(translatedHtmlElements.some(
+            element => element === originalElement)) ||
+          originalHtmlElements.filter(
+            element => element === originalElement).length >
+            translatedHtmlElements.filter(
+              element => element === originalElement).length
+        ) {
+          return false;
+        }
+      }
```

Comment:
The element arrays should be ordered right? Can we compare index by index? Something like

```
if (originalElementTagNames.length !== translatedElementTagNames) return false;

for (const [i, originalTagName] of originalHtmlElements.entries()) {
  if (originalTagName !== translatedElementTagNames[i]) return false;
}
```